fix: Don't wait for output post-processor in test suite

mclasmeier opened the pull request:
This is spawned from ROX-30713.

## tl,dr

This PR removes the subprocess waiting in the run-scanner-v4-install test suite to properly terminate on command failures.

## Longer Description

There's a problem in the run-scanner-v4-install test suite currently: When a command fails, the test case
will simply block instead of terminating with an error. This could hide important logs on failures.

For example, when adding this test test on master:

```
@test "Test" {
    init
    _begin "foo"
    echo A
    echo B
    false
    echo C
}
```

the test suite will progress as far as:

```
[...]
INFO: Wed Oct  1 10:08:27 CEST 2025: [pre-test-tear-down] Wed Oct  1 10:08:27 CEST 2025: Finished tearing down resources.
INFO: Wed Oct  1 10:08:27 CEST 2025: [pre-test-tear-down] Finished test setup.
INFO: Wed Oct  1 10:08:27 CEST 2025: [pre-test-tear-down] TIMING_DATA: {"test": "Test", "step": "pre-test-tear-down", "seconds_spent": 16}
INFO: Wed Oct  1 10:08:27 CEST 2025: [foo] A
INFO: Wed Oct  1 10:08:27 CEST 2025: [foo] B
```

and then sit there until force-killed.

With the fix it will terminate as expected:
```
INFO: Wed Oct  1 10:02:51 CEST 2025: [pre-test-tear-down] Wed Oct  1 10:02:51 CEST 2025: Finished tearing down resources.
INFO: Wed Oct  1 10:02:51 CEST 2025: [pre-test-tear-down] Finished test setup.
INFO: Wed Oct  1 10:02:51 CEST 2025: [pre-test-tear-down] TIMING_DATA: {"test": "Test", "step": "pre-test-tear-down", "seconds_spent": 1}
INFO: Wed Oct  1 10:02:51 CEST 2025: [foo] A
INFO: Wed Oct  1 10:02:51 CEST 2025: [foo] B
Aborting due to test failure.
   (in test file tests/e2e/run-scanner-v4-install.bats, line 451)
INFO: Wed Oct  1 10:02:51 CEST 2025: [post-test-tear-down] FAILED: Test "Test" failed. Look above for the test steps that have led to this failure.
     `false' failed
   TIMING_DATA: {"test": "Test", "step": "foo", "seconds_spent": 0}
```

Explanation of the fix: While it seems reasonable to wait for the output post-processor spawned earlier, it
1. should not be required, because the post-processor terminates by itself when stdin is at EOF.
2. doesn't play nice with bats, because bats does some wild hackery with file descriptors and process trees, which seems to be causing the `wait` to
effectively wait for the wrong thing. At least, Claude convinced me that this is the reason and at least it is consistent with the observations.

## User-facing documentation & Tests

Nothing user facing, just improving a test suite.

### How I validated my change

Manual testing, see above in the longer description.

porridge commented:
I also did a run with a failing test in other places relative to other tests, and could not reproduce the hang either 🤷🏻 
https://github.com/stackrox/stackrox/pull/17143

mclasmeier commented:
> While this should help with the hang, I'm not sure this will help with the missing test output.
> 
> I'm not buying the claude explanation about mysterious things happening.
> 
> Initially I thought this could hang if we mistakenly invoke the arg-less form of `wait` - where it waits for **everything**. But later I realized that since the arg is quoted, we'd never get there, we'd get something like this instead:
> 
> ```
> [stackrox]$ wait ""
> -bash: wait: `': not a pid or valid job spec
> [stackrox]$ 
> ```
> 
> So what must be happening is that the output processor process is deadlocked, and thus never exits. And this could happen if if blocks on a write, while BATS is not reading, for whatever reason. And if BATS is not reading, then this will not help with the missing output :-(
> 
> BTW, I made [this modification](https://github.com/stackrox/stackrox/pull/17142/files) on master and got the following result:
> 
> ```
> INFO: Mon Oct 6 11:14:28 UTC 2025: [setup-test-env] Executing Test: Test crash
> INFO: Mon Oct 6 11:14:28 UTC 2025: [setup-test-env] Mon Oct 6 11:14:28 UTC 2025: Setting up GCP auth and config
> INFO: Mon Oct 6 11:14:29 UTC 2025: [setup-test-env] Current project is already set to acs-san-stackroxci. Assuming configuration already applied.
> INFO: Mon Oct 6 11:14:29 UTC 2025: [setup-test-env] GOOGLE_APPLICATION_CREDENTIALS: /tmp/gcp.json
> INFO: Mon Oct 6 11:14:29 UTC 2025: [setup-test-env] Mon Oct 6 11:14:29 UTC 2025: Setting up the deployment environment
> INFO: Mon Oct 6 11:14:29 UTC 2025: [setup-test-env] TIMING_DATA: {"test": "Test_crash", "step": "setup-test-env", "seconds_spent": 1}
> INFO: Mon Oct 6 11:14:29 UTC 2025: [pre-test-tear-down] Mon Oct 6 11:14:29 UTC 2025: Will remove any existing stackrox resources
> INFO: Mon Oct 6 11:14:29 UTC 2025: [pre-test-tear-down] Mon Oct 6 11:14:29 UTC 2025: Tearing down StackRox resources for namespaces stackrox-central stackrox-sensor stackrox...
> INFO: Mon Oct 6 11:14:31 UTC 2025: [pre-test-tear-down] out: No resources found
> INFO: Mon Oct 6 11:14:31 UTC 2025: [pre-test-tear-down] out: Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "securitypolicies.config.stackrox.io" not found
> INFO: Mon Oct 6 11:14:31 UTC 2025: [pre-test-tear-down] Mon Oct 6 11:14:31 UTC 2025: Finished tearing down resources.
> INFO: Mon Oct 6 11:14:31 UTC 2025: [pre-test-tear-down] Finished test setup.
> INFO: Mon Oct 6 11:14:31 UTC 2025: [pre-test-tear-down] TIMING_DATA: {"test": "Test_crash", "step": "pre-test-tear-down", "seconds_spent": 2}
> INFO: Mon Oct 6 11:14:31 UTC 2025: [test-crash] A
> INFO: Mon Oct 6 11:14:31 UTC 2025: [test-crash] B
> INFO: Mon Oct 6 11:14:31 UTC 2025: [post-test-tear-down] FAILED: Test "Test crash" failed. Look above for the test steps that have led to this failure.
> INFO: Mon Oct 6 11:14:31 UTC 2025: [post-test-tear-down] Attempting to collect analysis data...
> INFO: Mon Oct 6 11:14:33 UTC 2025: [post-test-tear-down] Post-test teardown complete.
> INFO: Mon Oct 6 11:14:33 UTC 2025: [post-test-tear-down] TIMING_DATA: {"test": "Test_crash", "step": "post-test-tear-down", "seconds_spent": 2}
> not ok 2 Test crash # in 5393 ms
> # (in test file tests/e2e/run-scanner-v4-install.bats, line 511)
> # `false' failed
> # TIMING_DATA: {"test": "Test_crash", "step": "test-crash", "seconds_spent": 0}
> # Last output:
> # INFO: Mon Oct 6 11:14:32 UTC 2025: Will remove any existing stackrox resources
> # INFO: Mon Oct 6 11:14:32 UTC 2025: Tearing down StackRox resources for namespaces stackrox-central stackrox-sensor stackrox...
> # out: No resources found
> # out: Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "securitypolicies.config.stackrox.io" not found
> # INFO: Mon Oct 6 11:14:33 UTC 2025: Finished tearing down resources.
> ```
> 
> I.e. no hang. How did you get your suite to hang? Where relative to other tests did you add the toy test?

For me this is pretty simple:

```
stackrox on  master [$!?] via 🐹 v1.24.3 via 🐍 v3.12.10 on ☁️  mclasmei@redhat.com 
❯ git diff
diff --git a/tests/e2e/run-scanner-v4-install.bats b/tests/e2e/run-scanner-v4-install.bats
index 9534c43d25..47e702e936 100755
--- a/tests/e2e/run-scanner-v4-install.bats
+++ b/tests/e2e/run-scanner-v4-install.bats
@@ -443,6 +443,15 @@ collect_analysis_data() {
 }
 export -f collect_analysis_data
 
+@test "Test crash" {
+    init
+    _begin "test-crash"
+    echo A
+    echo B
+    false
+    echo C
+}
+
 @test "Upgrade from old Helm chart to HEAD Helm chart with Scanner v4 enabled" {
     init
 
```

and then
```
stackrox on  master [$!?] via 🐹 v1.24.3 via 🐍 v3.12.10 on ☁️  mclasmei@redhat.com 
❮ ORCHESTRATOR_FLAVOR=k8s BATS_CORE_ROOT=$HOME/bats-core bats  ./tests/e2e/run-scanner-v4-install.bats
run-scanner-v4-install.bats

INFO: Tue Oct  7 09:18:03 CEST 2025: [setup-file]     _    ____ ____    ___           _        _ _       _   _               _____         _
INFO: Tue Oct  7 09:18:03 CEST 2025: [setup-file]    / \  / ___/ ___|  |_ _|_ __  ___| |_ __ _| | | __ _| |_(_) ___  _ __   |_   _|__  ___| |_ ___
INFO: Tue Oct  7 09:18:03 CEST 2025: [setup-file]   / _ \| |   \___ \   | || '_ \/ __| __/ _` | | |/ _` | __| |/ _ \| '_ \    | |/ _ \/ __| __/ __|
INFO: Tue Oct  7 09:18:03 CEST 2025: [setup-file]  / ___ \ |___ ___) |  | || | | \__ \ || (_| | | | (_| | |_| | (_) | | | |   | |  __/\__ \ |_\__ \
INFO: Tue Oct  7 09:18:03 CEST 2025: [setup-file] /_/   \_\____|____/  |___|_| |_|___/\__\__,_|_|_|\__,_|\__|_|\___/|_| |_|   |_|\___||___/\__|___/
INFO: Tue Oct  7 09:18:03 CEST 2025: [setup-file] 
INFO: Tue Oct  7 09:18:03 CEST 2025: [setup-file] warning: tag '4.9' is externally known as '4.9.x'
INFO: Tue Oct  7 09:18:07 CEST 2025: [setup-file] warning: tag '4.9' is externally known as '4.9.x'
INFO: Tue Oct  7 09:18:07 CEST 2025: [setup-file] Tue Oct  7 09:18:07 CEST 2025: Using MAIN_IMAGE_TAG=4.9.x-992-gfdffcae437-dirty
INFO: Tue Oct  7 09:18:07 CEST 2025: [setup-file] Cloning released Helm charts into /tmp/helm-charts.iXJpBw...
INFO: Tue Oct  7 09:18:13 CEST 2025: [setup-file] Tue Oct  7 09:18:13 CEST 2025: Using EARLIER_CHART_VERSION=4.8.4
INFO: Tue Oct  7 09:18:13 CEST 2025: [setup-file] Tue Oct  7 09:18:13 CEST 2025: roxctl is built in non-release mode.
INFO: Tue Oct  7 09:18:13 CEST 2025: [setup-file] Tue Oct  7 09:18:13 CEST 2025: Local modification of the Helm charts will be taken into account for Helm chart tests.
INFO: Tue Oct  7 09:18:13 CEST 2025: [setup-file] Tue Oct  7 09:18:13 CEST 2025: MAIN_IMAGE_TAG (4.9.x-992-gfdffcae437-dirty) does not match roxctl version (4.9.x-772-g9b7db6b3e6).
INFO: Tue Oct  7 09:18:13 CEST 2025: [setup-file] For tests requiring an older roxctl version, /Users/mclasmeier/go/src/github.com/stackrox/stackrox/old-roxctl will be used.
INFO: Tue Oct  7 09:18:13 CEST 2025: [setup-file] Using OPERATOR_VERSION_TAG=4.9.0-992-gfdffcae437-dirty
INFO: Tue Oct  7 09:18:13 CEST 2025: [setup-file] Tue Oct  7 09:18:13 CEST 2025: Setting up default certs for tests
INFO: Tue Oct  7 09:18:13 CEST 2025: [setup-file] /var/folders/mt/4vkmn4ps3dq_ylx6drvnl1ph0000gn/T/tmp.c2RDQgtVTB ~/go/src/github.com/stackrox/stackrox
INFO: Tue Oct  7 09:18:13 CEST 2025: [setup-file] -----
INFO: Tue Oct  7 09:18:14 CEST 2025: [setup-file] Certificate request self-signature ok
INFO: Tue Oct  7 09:18:14 CEST 2025: [setup-file] subject=CN=Intermediate Server CA
INFO: Tue Oct  7 09:18:14 CEST 2025: [setup-file] Certificate request self-signature ok
INFO: Tue Oct  7 09:18:14 CEST 2025: [setup-file] subject=CN=custom-tls-cert.central.stackrox.local
INFO: Tue Oct  7 09:18:14 CEST 2025: [setup-file] ~/go/src/github.com/stackrox/stackrox
INFO: Tue Oct  7 09:18:14 CEST 2025: [setup-file] Contents of /var/folders/mt/4vkmn4ps3dq_ylx6drvnl1ph0000gn/T/tmp.c2RDQgtVTB:
INFO: Tue Oct  7 09:18:14 CEST 2025: [setup-file] total 52
INFO: Tue Oct  7 09:18:14 CEST 2025: [setup-file] drwx------  13 mclasmeier staff  416 Oct  7 09:18 .
INFO: Tue Oct  7 09:18:14 CEST 2025: [setup-file] drwx------ 221 mclasmeier staff 7072 Oct  7 09:18 ..
INFO: Tue Oct  7 09:18:14 CEST 2025: [setup-file] -rw-r--r--   1 mclasmeier staff 1082 Oct  7 09:18 ca.crt
INFO: Tue Oct  7 09:18:14 CEST 2025: [setup-file] -rw-------   1 mclasmeier staff 1704 Oct  7 09:18 ca.key
INFO: Tue Oct  7 09:18:14 CEST 2025: [setup-file] -rw-r--r--   1 mclasmeier staff   41 Oct  7 09:18 ca.srl
INFO: Tue Oct  7 09:18:14 CEST 2025: [setup-file] -rw-r--r--   1 mclasmeier staff 1509 Oct  7 09:18 intermediate.crt
INFO: Tue Oct  7 09:18:14 CEST 2025: [setup-file] -rw-------   1 mclasmeier staff 3268 Oct  7 09:18 intermediate.key
INFO: Tue Oct  7 09:18:14 CEST 2025: [setup-file] -rw-r--r--   1 mclasmeier staff   41 Oct  7 09:18 intermediate.srl
INFO: Tue Oct  7 09:18:14 CEST 2025: [setup-file] -rw-------   1 mclasmeier staff 6211 Oct  7 09:18 keystore.p12
INFO: Tue Oct  7 09:18:14 CEST 2025: [setup-file] -rw-r--r--   1 mclasmeier staff 1911 Oct  7 09:18 leaf.crt
INFO: Tue Oct  7 09:18:14 CEST 2025: [setup-file] -rw-------   1 mclasmeier staff 3272 Oct  7 09:18 leaf.key
INFO: Tue Oct  7 09:18:14 CEST 2025: [setup-file] -rw-r--r--   1 mclasmeier staff 4502 Oct  7 09:18 tls.crt
INFO: Tue Oct  7 09:18:14 CEST 2025: [setup-file] -rw-------   1 mclasmeier staff 3272 Oct  7 09:18 tls.key
INFO: Tue Oct  7 09:18:14 CEST 2025: [setup-file] Rendering fresh central-services Helm chart and writing to /var/folders/mt/4vkmn4ps3dq_ylx6drvnl1ph0000gn/T/tmp.IXkBuGsZnu...
INFO: Tue Oct  7 09:18:14 CEST 2025: [setup-file] WARN: Removed output directory /var/folders/mt/4vkmn4ps3dq_ylx6drvnl1ph0000gn/T/tmp.IXkBuGsZnu
INFO: Tue Oct  7 09:18:14 CEST 2025: [setup-file] INFO: Written Helm chart central-services to directory "/var/folders/mt/4vkmn4ps3dq_ylx6drvnl1ph0000gn/T/tmp.IXkBuGsZnu"
INFO: Tue Oct  7 09:18:14 CEST 2025: [setup-file] Rendering fresh secured-cluster-services Helm chart and writing to /var/folders/mt/4vkmn4ps3dq_ylx6drvnl1ph0000gn/T/tmp.9SisAYaD9V...
INFO: Tue Oct  7 09:18:14 CEST 2025: [setup-file] WARN: Removed output directory /var/folders/mt/4vkmn4ps3dq_ylx6drvnl1ph0000gn/T/tmp.9SisAYaD9V
INFO: Tue Oct  7 09:18:14 CEST 2025: [setup-file] INFO: Written Helm chart secured-cluster-services to directory "/var/folders/mt/4vkmn4ps3dq_ylx6drvnl1ph0000gn/T/tmp.9SisAYaD9V"
INFO: Tue Oct  7 09:18:14 CEST 2025: [setup-file] Patching node kind-cluster-control-plane to include label run-collector=true
INFO: Tue Oct  7 09:18:14 CEST 2025: [setup-file] node/kind-cluster-control-plane not labeled
INFO: Tue Oct  7 09:18:14 CEST 2025: [setup-file] collector will only be scheduled on node kind-cluster-control-plane
INFO: Tue Oct  7 09:18:14 CEST 2025: [setup-file] TIMING_DATA: {"test": "", "step": "setup-file", "seconds_spent": 11}
 ✗ Test crash
INFO: Tue Oct  7 09:18:15 CEST 2025: [setup-test-env] Executing Test: Test crash
INFO: Tue Oct  7 09:18:15 CEST 2025: [setup-test-env] TIMING_DATA: {"test": "Test_crash", "step": "setup-test-env", "seconds_spent": 0}
INFO: Tue Oct  7 09:18:15 CEST 2025: [pre-test-tear-down] Tue Oct  7 09:18:15 CEST 2025: Will remove any existing stackrox resources
INFO: Tue Oct  7 09:18:15 CEST 2025: [pre-test-tear-down] Tue Oct  7 09:18:15 CEST 2025: Tearing down StackRox resources for namespaces stackrox-central stackrox-sensor stackrox...
INFO: Tue Oct  7 09:18:15 CEST 2025: [pre-test-tear-down] out: No resources found
INFO: Tue Oct  7 09:18:15 CEST 2025: [pre-test-tear-down] out: Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "securitypolicies.config.stackrox.io" not found
INFO: Tue Oct  7 09:18:15 CEST 2025: [pre-test-tear-down] Tue Oct  7 09:18:15 CEST 2025: Finished tearing down resources.
INFO: Tue Oct  7 09:18:15 CEST 2025: [pre-test-tear-down] Finished test setup.
INFO: Tue Oct  7 09:18:15 CEST 2025: [pre-test-tear-down] TIMING_DATA: {"test": "Test_crash", "step": "pre-test-tear-down", "seconds_spent": 0}
INFO: Tue Oct  7 09:18:16 CEST 2025: [test-crash] A
INFO: Tue Oct  7 09:18:16 CEST 2025: [test-crash] B
```

Need to terminate with ^C.

porridge commented:
Thanks for the instructions @mclasmeier - I can also reproduce this locally.
I guess on CI is different, possibly because there's no controlling terminal.

I took a look at the process table during the hang on `wait`:
```
[stackrox]$ ps axfu | grep [p]ost_process_output
mowsiany   54984  0.0  0.0   4272  3148 pts/0    S+   09:28   0:00  |   |       |               \_ bash -c post_process_output 'test-crash'
```

Let's see which pipe it has open for reading:

```
[stackrox]$ ls -l /proc/54984/fd|grep ' 0 -> '
lr-x------ 1 4200660 4200660 64 Oct  7 09:32 0 -> pipe:[1371774]
```

OK, let's see if anyone has it open for writing:

```
[stackrox]$ find /proc/*/fd -ls 2>/dev/null|grep 1371774
  1385588      0 l-wx------   1 4200660  4200660               64 Oct  7 09:32 /proc/54327/fd/13 -> pipe:[1371774]
  1385589      0 l-wx------   1 4200660  4200660               64 Oct  7 09:32 /proc/54327/fd/14 -> pipe:[1371774]
  1385597      0 lr-x------   1 4200660  4200660               64 Oct  7 09:32 /proc/54982/fd/0 -> pipe:[1371774]
  1385606      0 lr-x------   1 4200660  4200660               64 Oct  7 09:32 /proc/54984/fd/0 -> pipe:[1371774]
[stackrox]$ 
```

Aha, PID 54327 has copied stdin/out into FDs 13 and 14... I guess this is some bats helper:

```
[stackrox]$ ps www 54327
    PID TTY      STAT   TIME COMMAND
  54327 pts/0    S+     0:00 /nix/store/izpf49b74i15pcr9708s3xdwyqs4jxwl-bash-5.2p32/bin/bash /nix/store/6b7n8pjhr4nblz93jcckd038fpcbwb3s-bats-1.11.0/libexec/bats-core/bats-exec-test --dummy-flag -x /home/mowsiany/go/src/github.com/stackrox/stackrox/tests/e2e/run-scanner-v4-install.bats test_Test_crash_first 1 1 1
[stackrox]$ 
```

So until this guy has the FD open, our postprocessor will not exit. However not waiting for it will be racy, and might cause lost output. 😟 

I tried a change where we try to synchronize with the processor by having the `_end` emit a special echo line to let the processor know it should exit. But this does not work, because, at turns out, the `_end` in this context (i.e. at least when the test fails due to command failure + errexit) does not have its stdout/err connected to the processor any more. They lead to the bats temporary output file.

So I guess we don't have much choice here. Let's give it a spin and see.
